wallet-new/src/Cardano/Wallet/Kernel/CoinSelection/Generic/Fees.hs

```diff
@@ -35,107 +36,167 @@ data FeeOptions dom = FeeOptions {
+            -- a change value causes an overflow).
+            --
+            -- Because selecting a new input increases the fee, we need to
+            -- re-run the algorithm with this new elements and using the initial
+            -- change plus the extra change brought up by this entry and see if
+            -- we can now correctly cover fee.
+        else do
+            extraUtxo <- coverRemainingFee pickUtxo remainingFee
+            let inps' = selectedEntries extraUtxo
+            let extraChange = selectedBalance extraUtxo
+            go (inps <> inps') outs (splitChange extraChange chgs)
+
+
+coverRemainingFee
+    :: forall utxo m. (Monad m, CoinSelDom (Dom utxo))
+    => PickUtxo m utxo
+    -> Fee (Dom utxo)
+    -> CoinSelT utxo CoinSelHardErr m (SelectedUtxo (Dom utxo))
 coverRemainingFee pickUtxo fee = go emptySelection
   where
     go :: SelectedUtxo (Dom utxo)
-       -> CoinSelT utxo CoinSelHardErr m (SelectedUtxo (Dom utxo), Value (Dom utxo))
+       -> CoinSelT utxo CoinSelHardErr m (SelectedUtxo (Dom utxo))
     go !acc
       | selectedBalance acc >= getFee fee =
-          return (acc, unsafeValueSub (selectedBalance acc) (getFee fee))
+          return acc
       | otherwise = do
           mio <- (pickUtxo $ unsafeValueSub (getFee fee) (selectedBalance acc))
           io  <- maybe (throwError CoinSelHardErrCannotCoverFee) return mio
           go (select io acc)
 
--- | Attempt to pay the fee from change outputs, returning any fee remaining
---
--- NOTE: For sender pays fees, distributing the fee proportionally over the
--- outputs is not strictly necessary (fairness is not a concern): we could just
--- use the change of the first output to cover the entire fee (if sufficiently
--- large). Doing it proportionally however has the benefit that the fee
--- adjustment doesn't change the payment:change ratio too much, which may be
--- important for the correct operation of the coin selection policy.
---
--- NOTE: This does mean that /if/ the policy generates small outputs with
--- very large corresponding change outputs, we may not make optional use of
--- those change outputs and perhaps unnecessarily add additional UTxO entries.
--- However, in most cases the policy cares about the output:change ratio,
--- so we stick with this approach nonetheless.
-feeFromChange :: forall dom. CoinSelDom dom
-              => Fee dom
-              -> [CoinSelResult dom]
-              -> ([CoinSelResult dom], Fee dom)
-feeFromChange totalFee =
-    bimap identity unsafeFeeSum
-    . unzip
-    . map go
-    . divvyFee (outVal . coinSelRequest) totalFee
-  where
-    -- | Adjust the change output, returning any fee remaining
-    go :: (Fee dom, CoinSelResult dom) -> (CoinSelResult dom, Fee dom)
-    go (fee, cs) =
-        let (change', fee') = reduceChangeOutputs fee (coinSelChange cs)
-        in (cs { coinSelChange = change' }, fee')
+
+-- we should have (x + (sum ls) = sum result), but this check could overflow.
```

Comment:
`splitChange` needs a comment what it does. Note that it makes no attempt to divvy the new output proportionally over the change outputs. This means that if we happen to pick a very large UTxO entry, adding this evenly rather than proportionally might skew the payment:change ratio a lot. Could consider defining this in terms of `divvy` instead (but if you do, we need to be wary of lower bounds/upper bounds in the division calculation, since the requirements may not be the same for _subtracting_ stuff and _adding_ stuff.). Can be reconsidered later.